### PR TITLE
feat: Introduce Optional field (diverseFromDeviceUuid )in create device - Single(non HA)

### PR DESCRIPTION
**Pull request by kpdhulipala**
feat: Introduce Optional field (diverseFromDeviceUuid )in create device - Single/non HA
Closes: ne-go module PR- https://github.com/equinix/ne-go/pull/31
https://developer.equinix.com/catalog/network-edgev1#operation/createVirtualDeviceUsingPOST

### CODE REVIEW

File: equinix/resource_network_device.go [hunk 1]
```diff
@@ -414,6 +418,20 @@ func createNetworkDeviceSchema() map[string]*schema.Schema {
...
+			Optional:      true,
+			ValidateFunc:  validation.IsUUID,
+			ConflictsWith: []string{neDeviceSchemaNames["Secondary"]},
+			Description:   neDeviceDescriptions["DiverseFromDeviceUUID"],
+		},
+		neDeviceSchemaNames["DiverseFromDeviceName"]: {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Optional:    true,
+			Description: neDeviceDescriptions["DiverseFromDeviceName"],
```
Comment (displague) [inline equinix/resource_network_device.go:433]: Does the device name diversity option also conflict with Secondary, as the UUID option does?

File: docs/resources/equinix_network_device.md [hunk 2]
```diff
@@ -298,6 +299,9 @@ device configurations. See [Secondary Device](#secondary-device) below for more
 If not specified, default will be INTERNET-ACCESS
 * `project_id` - (Optional) Unique Identifier for the project resource where the device is scoped to.If you
 leave it out, the device will be created under the default project id of your organization.
+* `diverse_device_id` - (Optional) Unique ID of an existing device.
+Use this field to let Equinix know if you want your new device to be in a different location from any existing virtual 
+device. This field is only meaningful for single devices.
```
Comment (displague) [inline docs/resources/equinix_network_device.md:305]: the diverse_device_name field is not documented

File: equinix/resource_network_device.go (hunk 1 above)
Comment (kpdhulipala) [reply equinix/resource_network_device.go:433]: no @displague , only diverse device uuid is being passed in the payload. diverse device name is present only in device details response(data source - network device)

File: docs/resources/equinix_network_device.md (hunk 2 above)
Comment (kpdhulipala) [reply docs/resources/equinix_network_device.md:307]: addressed them, pls check and let me know @displague

File: equinix/resource_network_device.go [hunk 3]
```diff
@@ -1062,6 +1080,12 @@ func createNetworkDevices(d *schema.ResourceData) (*ne.Device, *ne.Device) {
 	if v, ok := d.GetOk(neDeviceSchemaNames["ProjectID"]); ok {
 		primary.ProjectID = ne.String(v.(string))
 	}
+	if v, ok := d.GetOk(neDeviceSchemaNames["DiverseFromDeviceUUID"]); ok {
+		primary.DiverseFromDeviceUUID = ne.String(v.(string))
+	}
+	if v, ok := d.GetOk(neDeviceSchemaNames["DiverseFromDeviceName"]); ok {
```
Comment (rling-equinix) [inline equinix/resource_network_device.go:1086]: This should be removed as we are not sending diverse device name in the request.

File: equinix/resource_network_device_test.go [hunk 4]
```diff
@@ -23,57 +23,60 @@ func TestNetworkDevice_createFromResourceData(t *testing.T) {
...
+		neDeviceSchemaNames["TermLength"]:            ne.IntValue(expectedPrimary.TermLength),
+		neDeviceSchemaNames["AdditionalBandwidth"]:   ne.IntValue(expectedPrimary.AdditionalBandwidth),
+		neDeviceSchemaNames["OrderReference"]:        ne.StringValue(expectedPrimary.OrderReference),
+		neDeviceSchemaNames["InterfaceCount"]:        ne.IntValue(expectedPrimary.InterfaceCount),
+		neDeviceSchemaNames["WanInterfaceId"]:        ne.StringValue(expectedPrimary.WanInterfaceId),
+		neDeviceSchemaNames["CoreCount"]:             ne.IntValue(expectedPrimary.CoreCount),
+		neDeviceSchemaNames["IsSelfManaged"]:         ne.BoolValue(expectedPrimary.IsSelfManaged),
+		neDeviceSchemaNames["ProjectID"]:             ne.StringValue(expectedPrimary.ProjectID),
+		neDeviceSchemaNames["DiverseFromDeviceUUID"]: ne.StringValue(expectedPrimary.DiverseFromDeviceUUID),
+		neDeviceSchemaNames["DiverseFromDeviceName"]: ne.StringValue(expectedPrimary.DiverseFromDeviceName),
```
Comment (rling-equinix) [inline equinix/resource_network_device_test.go:79]: This should also be removed.

**Review by rling-equinix — CHANGES_REQUESTED** [review]
Could you please add the test cases for data source network device?

File: equinix/resource_network_device_test.go (hunk 4 above)
Comment (kpdhulipala) [reply equinix/resource_network_device_test.go]: removed

File: equinix/resource_network_device.go (hunk 3 above)
Comment (kpdhulipala) [reply equinix/resource_network_device.go]: removed

File: equinix/resource_network_device.go [hunk 5]
```diff
@@ -414,6 +418,20 @@ func createNetworkDeviceSchema() map[string]*schema.Schema {
...
+			Computed:      true,
+			Optional:      true,
+			ValidateFunc:  validation.IsUUID,
+			ConflictsWith: []string{neDeviceSchemaNames["Secondary"]},
+			Description:   neDeviceDescriptions["DiverseFromDeviceUUID"],
+		},
+		neDeviceSchemaNames["DiverseFromDeviceName"]: {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Optional:    true,
```
[the remark this answers is not in the document]
Comment (kpdhulipala) [reply equinix/resource_network_device.go]: addressed

File: docs/resources/equinix_network_device.md [hunk 6]
```diff
@@ -298,6 +299,11 @@ device configurations. See [Secondary Device](#secondary-device) below for more
...
+* `diverse_device_name` - Name of the device with diverse device uuid. This field is returned in device details if the
```
[the remark this answers is not in the document]
Comment (kpdhulipala) [reply docs/resources/equinix_network_device.md]: updated in attributes reference section

**Comment by kpdhulipala** [comment]
> Could you please add the test cases for data source network device?

added one test case for data source network device, will add more in coming PRs

**Review by rling-equinix — APPROVED** [review]
Lgtm

File: docs/resources/equinix_network_device.md (hunk 2 above)
Comment (displague) [reply docs/resources/equinix_network_device.md:305]: The fields should be added here in the datasource documentation:
https://github.com/kpdhulipala/terraform-provider-equinix/blame/NFV-27633/docs/data-sources/equinix_network_device.md#L31

Comment (kpdhulipala) [reply docs/resources/equinix_network_device.md:305]: sorry, I got it now, added them in data source equinix_network_device.md

**Review by displague — APPROVED** [review]

**Pull request merged**